### PR TITLE
Promote minReadySeconds to GA

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Promote StatefulSet minReadySeconds to GA. This means `--feature-gates=StatefulSetMinReadySeconds=true` are not needed on kube-apiserver and kube-controller-manager binaries and they'll be removed soon following policy at https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: <link> https://github.com/kubernetes/enhancements/pull/3354 
- [Usage]: Promote StatefulSet minReadySeconds to GA. This means `--feature-gates=StatefulSetMinReadySeconds=true` are not needed on kube-apiserver and kube-controller-manager binaries and they'll be removed soon following policy at https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation
- [Usage]: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#minimum-ready-seconds

```

### CODE REVIEW

**Comment by leilajal** [comment]
/assign @alexzielenski @yliaog 
thanks for looking into this!
/triage accepted

**Review by alexzielenski — APPROVED** [review]
/lgtm

**Comment by ravisantoshgudimetla** [comment]
/assign @liggitt

for API reviews

File: pkg/apis/apps/types.go [hunk 1]
```diff
@@ -206,7 +206,6 @@ type StatefulSetSpec struct {
 	// Minimum number of seconds for which a newly created pod should be ready
 	// without any of its container crashing for it to be considered available.
 	// Defaults to 0 (pod will be considered available as soon as it is ready)
-	// This is an alpha field and requires enabling StatefulSetMinReadySeconds feature gate.
```
Comment (ravisantoshgudimetla) [inline pkg/apis/apps/types.go:209]: @liggitt - Should I open backports to make this beta in previous releases?

**Comment by liggitt** [comment]
API changes lgtm. The question about the TODO in pkg/controller/statefulset/stateful_set_control.go needs answering

**Comment by liggitt** [comment]
Needs a more descriptive release note that would make sense to API consumers (for the API field) and kube-apiserver administrators (for the feature gate GA and eventual removal)

File: pkg/controller/statefulset/stateful_set_control.go [hunk 2]
```diff
@@ -462,9 +458,7 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 		// If we have a Pod that has been created but is not available we can not make progress.
 		// We must ensure that all for each Pod, when we create it, all of its predecessors, with respect to its
 		// ordinal, are Available.
-		// TODO: Since available is superset of Ready, once we have this featuregate enabled by default, we can remove the
-		// isRunningAndReady block as only Available pods should be brought down.
-		if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetMinReadySeconds) && !isRunningAndAvailable(replicas[i], set.Spec.MinReadySeconds) && monotonic {
+		if !isRunningAndAvailable(replicas[i], set.Spec.MinReadySeconds) && monotonic {
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply pkg/controller/statefulset/stateful_set_control.go:461]: Yes, I initially added this TODO of removing it as `Available` is a superset of `Ready` but thinking more about it, I felt it is a bit unreadable to combine both. I don't have strong opinions one way or the other.

[the remark this answers is not in the document]
Comment (atiratree) [reply pkg/controller/statefulset/stateful_set_control.go:461]: I am okay with keeping both check for readability

[the remark this answers is not in the document]
Comment (krmayankk) [reply pkg/controller/statefulset/stateful_set_control.go:461]: +1 for not combining ready and available. Also the error messages provide extra context and differentiate between ready and available

**Comment by liggitt** [comment]
alpha failure does not look related to this gate

/retest

File: pkg/apis/apps/types.go (hunk 1 above)
Comment (liggitt) [reply pkg/apis/apps/types.go:209]: probably so, since the API docs are incorrect

**Comment by liggitt** [comment]
one nit on the feature gate removal comment, and would like an approve from sig-apps-approvers

**Comment by liggitt** [comment]
/approve
for API changes

needs an approve from an apps approver

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
from sig-apps pov

**Pull request merged**